### ISSUE TITLE
how to start training？

### ISSUE THREAD

**Issue by Brightlcz** [lpcccc-cv/CFFNet#2]
Dr. Lei, thank you for your excellent medical slice interpolation paper!

The code test is normal, but how to create the training data? Is it using create_lmdb_mp_HR.py to create the training data set? and what exactly does train.txt look like?

Looking forward to your reply.

**Comment by lpcccc-cv** [comment]
For the single slice interpolation, there are three images in each sample. The folder structure is:
```
├── sequences
    ├── volume001
        ├── sample00001
            ├── im1.png
            ├── im2.png
            └── im3.png
        ├── sample00002
            ├── im1.png
            ├── im2.png
            └── im3.png ...
    ├── volume002
        ├── sample00001
            ├── im1.png
            ├── im2.png
            └── im3.png
        ├── sample00002
            ├── im1.png
            ├── im2.png
            └── im3.png ...
        ...
```
Then you can use  the create_lmdb_mp_HR.py to create the lmdb file for model training.
You can also refer to the Vimeo90k dataset, our data format is the same as it.

**Comment by Brightlcz** [comment]
Thank you very much for your reply, Dr. Lei. Thank you, brother. I will try it.

**Comment by Brightlcz** [comment]
Dr. Lei, I would like to ask you again. how to perform self-supervised training. It seems that changing ssl in the configuration file to True does not work. There is no optimize_parameters_self_supervise() function.

**Comment by lpcccc-cv** [comment]
For the SSL training. 
(1) You should prepare two datasets for the two-stage training;
(2) In the first stage, the training is same with the FSL training;
(3) In the second stage training, its differences are loss functions and learning rate.
About the datasets and the training details, you can find in our paper.

**Comment by lpcccc-cv** [comment]
You can refer to the following codes:

```
def optimize_parameters_self_supervise(self, step):
        self.optimizer_G.zero_grad()

        # pre-train result
        self.pretrain_fake_1 = self.netG_pretrain(self.var_L[:, [0, 1], :, :, :], t=self.var_T)
        self.pretrain_fake_2 = self.netG_pretrain(self.var_L[:, [1, 2], :, :, :], t=self.var_T)
        
        self.fake_H_1 = self.netG(self.var_L[:, [0, 1], :, :, :], t=self.var_T)
        self.fake_H_2 = self.netG(self.var_L[:, [1, 2], :, :, :], t=self.var_T)
        # print('fake_H_1.shape', self.fake_H_1.shape)
        
        self.real_H = self.var_L[:, 1, :, :, :]  # 输入三帧的中间帧始终为target
        b, n ,c, h, w = self.fake_H_1.size()   # n=9
        
        # cycle loss
        # cur_loss_1 = self.cri_pix(self.real_H, cur_output[:, 0, :, :, :])
        # slice_loss
        # cur_loss_2 = self.cri_pix(self.real_H, self.fake_H_0[:, 0, :, :, :])
        # smooth loss
        # l_smooth = self.cri_pix(self.pretrain_fake_1.detach(), self.fake_H_1) + self.cri_pix(self.pretrain_fake_2.detach(), self.fake_H_2)
        # print(l_smooth.item())
        # vgg loss
        # vgg_fea1, vgg_fea2 = self.vgg19(self.real_H.repeat(1, 3, 1, 1), supervised_output.repeat(1, 3, 1, 1))
        # vgg_loss = self.cri_pix(vgg_fea1, vgg_fea2)

        # 对fake_H滑动窗口,滑动次数为插入帧数
        total_loss = 0
        l_pix = 0
        # print(n)

        # for i in range(n):
        #随机生成i
        i = int(np.random.randint(0,5,size=1))
        cur_input = torch.stack((self.fake_H_1[:, i, :, :, :], self.fake_H_2[:, i, :, :, :]), 1)
        # print('cur_input', cur_input.shape)
        cur_output = self.netG(cur_input, t=self.var_T)
        # print(cur_output.shape)
        supervised_output = cur_output[:, -1*(i+1), :, :, :] 
        cur_loss = self.cri_pix(self.real_H, supervised_output)  
        # print(cur_loss)    
        # l_smooth = self.cri_pix(self.pretrain_fake_1[:, i, :, :, :].detach(), self.fake_H_1[:, i, :, :, :]) + self.cri_pix(self.pretrain_fake_2[:, i, :, :, :].detach(), self.fake_H_2[:, i, :, :, :])   
        l_smooth = self.vgg19(self.pretrain_fake_1[:, i, :, :, :].repeat(1, 3, 1, 1).detach(), self.fake_H_1[:, i, :, :, :].repeat(1, 3, 1, 1)) + self.vgg19(self.pretrain_fake_2[:, i, :, :, :].repeat(1, 3, 1, 1).detach(), self.fake_H_2[:, i, :, :, :].repeat(1, 3, 1, 1))
        l_pix = l_pix + cur_loss
        # total_loss = l_pix+l_smooth*0.003
        total_loss = l_pix
        total_loss.backward()
        self.optimizer_G.step()
        # set log
        self.log_dict['l_pix'] = l_pix.item()
        # self.log_dict['l_smooth'] = l_smooth.item()
        # self.log_dict['l_pix2'] = cur_loss_2.item()
        # self.log_dict['l_vgg'] = vgg_loss.item()  
```
